design/one-pager-consuming-k8s-clusters.md

```diff
@@ -0,0 +1,270 @@
+  managed Kubernetes service (e.g. GKE, AKS, EKS).
+* `KubernetesCluster` claim: a namespace-scoped custom resource that can be used
+  to claim a Kubernetes cluster managed resource, or dynamically provision one.
+* **Workload**: a collection of Kubernetes resources that are intended to be
+  deployed into a target Kubernetes cluster. A workload is represented by a
+  `KubernetesApplication` custom resource.
+
+## Background
+
+Currently, Crossplane models portable cloud provider managed services as
+*managed resources* that can be either statically or dynamically provisioned
+using resource *classes* and *claims*. One such resource type is a Kubernetes
+cluster, which uses a `KubernetesCluster` claim and cloud provider-specific
+managed resource types (e.g. `GKECluster`, `AKSCluster`, `EKSCluster`, etc.). It
+is appropriate for Kubernetes clusters to be modeled in the same manner as other
+managed services due to the fact that they are presented by the provider in the
+same fashion. However, in Crossplane, Kubernetes clusters are unique to other
+managed service offerings in that they can also be used as a target for running
+workloads.
+
+Workloads are provisioned using the `KubernetesApplication` type. It is
+namespace-scoped and is scheduled to a target cluster using labels to match a
+`KubernetesCluster` claim in its namespace. Because managed resources may only
+be bound to a single claim, the claim is namespaced, and `KubernetesApplication`
+resources can only be scheduled to `KubernetesCluster` claims in their
+namespace, any Kubernetes managed resource offering may only be consumed (i.e.
+deployed to) from a single namespace.
+
+In order to fully conceptualize the usage of Kubernetes clusters in Crossplane,
+it is helpful to separate the notions of a **Kubernetes cluster as a managed
+service** and a **Kubernetes cluster as a target for deploying workloads**. With
+this distinction made, we can assert that this proposal in no way modifies the
+managed service notion of a Kubernetes cluster, but instead formalizes this
+concept of a Kubernetes cluster as a deployment target.
+
+## Goals
+
+Because it is unlikely that Crossplane users will always want a 1-to-1
+namespace-to-cluster consumption model, it is desirable to support deploying
+workloads in any namespace to any Kubernetes cluster. However, it is also
+desirable to continue to treat managed Kubernetes services as a managed resource
+that is supplied by cloud providers. For this reason, any solution should be
+additive to the existing model.
+
+As a secondary goal for this design, it should be possible to use existing
+Kubernetes clusters (wherever they may be running) that were not provisioned by
+Crossplane as targets for deploying workloads. This concept has frequently been
+referred to as a "Bring Your Own Cluster" scenario.
+
+## Proposal
+
+This proposal can be separated into two main categories:
+1. The implementation of a new `WorkloadTarget` custom resource.
```

Comment:
Given that `KubernetesApplication` is just one kind of workload, I think `KubernetesApplicationTarget` would fit better here, though I don't love either name. `KubernetesCluster` feels like the best name for this, but it's (kind of) taken by the `KubernetesCluster` resource claim. We could have `kubernetescluster.workload.crossplane.io` as the scheduling target
and `kubernetescluster.compute.crossplane.io` as the claim, but that would probably be pretty confusing.

Reply:
Based on my current understanding, a draft of naming that makes sense in my mind:

`KubernetesClusterClaim` - For claiming a (workload) target cluster (currently `KubernetesCluster`) assuming here that kubernetes clusters can only be claimed to run workloads on, which is portable between cloud managed and custom/on-prem clusters.
`KubernetesCluster` - A workload target kubernetes cluster refering to managed or on-prem cluster (`WorkloadTarget` in this doc, since it only targets kubernetes)
`ManagedKubernetesCluster` - What is currently `KubernetesCluster` (which i believe only handles managed clusters, which I assume includes things such as scaling and upgrading their nodes, etc...)

I believe that a custom/on-prem workload target cluster should either be a separately managed `KubernetesCluster` (as per above naming,) or something like a `CustomKubernetesCluster` (where both `CustomKubernetesCluster` and `ManagedKubernetesCluster` reference or own the published `KubernetesCluster` workload target)

Will add some more notes about my thoughts on `Provider(s)` in a later comment.

Please let me know if I'm getting anything mixed up or any other clarficiations for my understanding would be appreciated!

Reply:
> `KubernetesClusterClaim` 

I definitely think it would make sense for our claim kinds to end with the word `Claim`, as our `Class` kinds end with the word `Class`. Though there is an interesting twist in Crossplane in that a `FooClaim` claims a `Bar`, not a `Foo`. 🤔

Reply:
> ManagedKubernetesCluster - What is currently KubernetesClusterResource? (which i believe only handles managed clusters, which I assume includes things such as scaling and upgrading their nodes, etc...)

No such kind exists today; are you thinking of our managed resources? They're "strongly typed" to their underlying cloud provider concept, so they're `kind: GKECluster`, `kind: EKSCluster` etc.

Reply:
> > ManagedKubernetesCluster - What is currently KubernetesClusterResource? (which i believe only handles managed clusters, which I assume includes things such as scaling and upgrading their nodes, etc...)
> 
> No such kind exists today; are you thinking of our managed resources? They're "strongly typed" to their underlying cloud provider concept, so they're `kind: GKECluster`, `kind: EKSCluster` etc.

Ok I see. Then I can just look at this `ManagedKubernetesCluster` as a theoretical abstraction of these cloud provider kinds. So if the cloud provider kinds are explicitly decoupled (and don't really have) a high level abstraction ("Managed" kind) by design, these names do make perfect sense to me.

Reply:
I would back either `KubernetesConnection` or `KubernetesTarget`. I think this makes sense as it seems like it may be true that we could schedule something other than a `KubernetesApplication` to a `KubernetesTarget/Connection`.

Reply:
I like that `KubernetesConnection` has symmetry with our existing "connection secret" concept.